clap_complete/src/dynamic/completer.rs

```diff
@@ -701,3 +720,96 @@ impl CompletionCandidate {
         self.hidden
     }
 }
+
+/// User-provided completion candidates for an argument.
+///
+/// This is useful when predefined value hints are not enough.
+pub trait CustomCompleter: core::fmt::Debug + Send + Sync {
+    /// All potential candidates for an argument.
+    ///
+    /// See [`CompletionCandidate`] for more information.
+    fn completions(&self) -> Vec<CompletionCandidate>;
+}
+
+/// A closure-based custom completer
+///
+/// The closure should return a list of completion candidates.
+pub struct ClosureCompleter<F>
+where
+    F: Fn() -> Vec<CompletionCandidate> + Send + Sync,
```

Comment:
The ergonomics of this aren't worth it.  Let's just hand impl `Debug` for `ArgValueCompleter` to just be the type name.